projects/web-wallet/rpc-protocol.md

```diff
@@ -0,0 +1,432 @@
+
+**Parameters**
+
+1. `token` - (Defaults to `HTR`) May be specified for a token other than `HTR`
+
+**Request:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "method": "htr_getBalance",
+  "params": {
+    "token": "000023a025d513c5c75410e771c297b23edc88b8fac7da1eff6fdc81e628c20d"
+  }
+}
+```
+**Response:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "result": {
+    "available": 500,
+    "locked": 0
+  }
+}
+```
+
+### 2.1.6 `htr_getConnectedNetwork`
+
+This method is used to allow the dApp to validate that it's running in the same network the wallet is connected to.
+
+**Parameters**
+
+**Request:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "method": "htr_getConnectedNetwork"
+}
+```
+**Response:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "result": {
+    "network": "mainnet" | "testnet" | "privatenet",
+    "genesisHash": "00000000000000000437c87c...c8bf9e9d2fda7de9763"
+  }
+}
+```
+
+### 2.1.7 `htr_getAddress`
+
+This method allows the dApps to request an address. The wallet implementation
+should decide the address to return
+
+**Parameters**
+
+**Request:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "method": "htr_getAddress"
+}
+```
+**Response:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "result": {
+    "address": "HNnK9wgUVL6Cjzs1K3jpoGgqQTXCqpAnW8"
+  }
+}
+```
+
+## 2.1.8 Error codes
+
+We should follow the [JSON-RPC](https://www.jsonrpc.org/specification#error_object) specification for error codes, so errors generated by the "server", which in our case, is the wallet handling the requests should be in the `-32099` to `-32000` range.
+
+| Error | Error Type | Description                        |
+|------------|------------------------------------|---|
+| -32050     | WALLET_NOT_READY | Wallet is not ready yet              |
+| -32051     | REQUEST_REJECTED | Request rejected by the user |
+| -32052     | MISSING_FUNDS | Missing funds to fulfill transaction |
+| -32053 | INVALID_PARAMETERS | Invalid parameters |
+| -32099     | GENERIC_FAILURE | Generic failure |
+
+# 3. Guide-level Explanation
+
+### 3.1 `htr_sendTx`
+
+When the request is received, we will validate if the parameters are valid, enforcing the data types and rules described in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+If the validation is successful, we will display a confirmation screen, with all outputs for manual user validation (UX is yet to be discussed). If the user rejects the transaction, we will respond with a `REQUEST_REJECTED` error
+
+If the user accepts the transaction, we will create a `SendTransaction` model (or `SendTransactionWalletService` if on the wallet-service) and execute the send transaction.
+
+Since `JSON-RPC` is designed considering only a single response per request, we will only respond the request when the transaction either failed to be pushed to the fullnode (or the wallet-service) or when it's successfully pushed.
```

Comment:
Should we have a `STILL WAITING` return code? In this case, the dApp could get information from a separate `/current-operation-status` API?

Reply:
I like this idea, I added a new RPC call to fetch the status for an operation (using its id) and added a specific `op_status` to the push tx RPC call, which is the one that has multiple states

b8eb722066bfa1edacf674589251c835bbd264cc


What do you think about this format?

Reply:
It seems great to me! Thanks!